### ISSUE TITLE
2.0 : Replicator does not resume after connection w/ Sync Gateway is temporarily broken

### ISSUE THREAD

**Issue by rajagp** [couchbase/couchbase-lite-ios#1978]
--------
* Version:   DB 20
* Client OS: iOS (Android has another issue)
* Server:    SGW 1.5.1

This may very well be  a Lite core issue but observed in iOS App. Could not test w/ Android due to a separate [issue](https://github.com/couchbase/couchbase-lite-android/issues/1471)
 
**Steps**:
Stop Sync Gateway 
Resume Sync Gateway
Try to sync some changes 
Replicator stuck in busy state

Reproducible using Travel Sample App

**Comment by rajagp** [comment]
On further investigation, when I stop the SGW , the `readClosedForStreamTask` callback on `NSURLSessionStreamDelegate`  is called. Thereafter, nothing happens on the socket

 This is the documentation on `readClosedForStreamTask`. Note the " but future reads will immediately fail" clause.
```
 /* Indiciates that the read side of a connection has been closed.  Any
 * outstanding reads complete, but future reads will immediately fail.
 * This may be sent even when no reads are in progress. However, when
 * this delegate message is received, there may still be bytes
 * available.  You only know that no more bytes are available when you
 * are able to read until EOF. */
```
`- (void)URLSession:(NSURLSession *)session readClosedForStreamTask:(NSURLSessionStreamTask *)streamTask;`

**Comment by snej** [comment]
Weird; why wouldn't the socket close completely if the remote process exits? I'll investigate.

**Comment by snej** [comment]
Sent a question to the macnetworkprog mailing list. I'll see if I can reproduce the problem and if so I'll experiment with adding code to this delegate method to close the connection.

**Comment by rajagp** [comment]
If it makes a difference , I have the SGW running in a docker container and I stop the container using `docker stop`.

**Comment by pasin** [comment]
I was able to reproduce this issue with Todo app. However I didn't see any delegate methods were called at all even `readClosedForStreamTask` after I shut down the SGW. The completion handler of `readDataOfMinLength:maxLength:timeout:completionHandler:` was not called as well even when the timeout was due.

Note: In 1.x, we use PSWebSocket in WS ChangeTracker, and we ended up need to implement read timeout to do retry by ourselves.

**Comment by pasin** [comment]
From the log, I did see the following and nothing else after shutting down SGW while the replicator is in IDLE state:

```
2018-01-18 20:51:39.372461-0800 Todo[87696:8905197] [Todo] Replicator: IDLE 1536/1536, error: (null)
2018-01-18 20:52:08.523428-0800 Todo[87696:8915221] TIC Read Status [2:0x608000161740]: 1:57
2018-01-18 20:52:08.523774-0800 Todo[87696:8915221] TIC TCP Conn Missing Error [2:0x608000161740]: Generating 1:54
```

**Comment by snej** [comment]
The "TIC…" log messages from CFNetwork(?) are the same ones I saw the other day when fixing the bogus ENOTCONN errors from closing the WebSocket connection. So in that case it does look like the socket close is detected.

**Comment by snej** [comment]
All that happens when the remote SG is killed is that the stream task's read completion handler gets called with the `atEOF` flag set. So I added code to close the stream. Then the `readClosedForStreamTask` delegate method is called. I modified that method to cancel the task ... and _then_ the tasks' `didCompleteWithError:` delegate method finally gets called.

I've now got CBLWebSocket reporting a POSIX ECONNRESET error back to LiteCore. This then triggers CBLReplication to retry the replicator since that error is considered transient.
